src/shims/unix/socket.rs

```diff
@@ -49,25 +57,23 @@ impl FileDescription for SocketPair {
         // need to be supported in the future, the check should be added here.
 
         let mut epoll_ready_events = EpollReadyEvents::new();
-        let readbuf = self.readbuf.borrow();
 
         // Check if it is readable.
+        let readbuf = self.readbuf.borrow();
         if !readbuf.buf.is_empty() {
             epoll_ready_events.epollin = true;
         }
 
         // Check if is writable.
-        if let Some(writebuf) = self.writebuf.upgrade() {
-            let writebuf = writebuf.borrow();
+        if let Some(peer_fd) = self.peer_fd().upgrade() {
+            let writebuf = &peer_fd.downcast::<SocketPair>().unwrap().readbuf.borrow();
             let data_size = writebuf.buf.len();
             let available_space = MAX_SOCKETPAIR_BUFFER_CAPACITY.strict_sub(data_size);
             if available_space != 0 {
                 epoll_ready_events.epollout = true;
             }
-        }
-
-        // Check if the peer_fd closed
-        if self.peer_fd.upgrade().is_none() {
```

Comment:
Okay looking at the old code here I think I have a theory for why behavior changed. But I can't imagine that the old behavior was intentional.

In think previously we had a situation where writebuf still existed but the peer was already destroyed. That's because a file descriptor reference gets invalidated *before* the call to `FileDescriptor::close`, but the `Rc` inside of it (in this case, for the buffer) survive until they get dropped, which happens at the *end of* `FileDescriptor::close`.